### ISSUE TITLE
props cause render even if I didn't use props in render

### ISSUE THREAD

**Issue by HealGaren** [mobxjs/mobx-react#543]
Hello.

I read #347 so I know mobx cannot react by part of props, But I thought an observer component's render function is "react" at all the time props change(filtered by shallow equality check).
In other words, I thought render function won't be called in an observer component that doesn't use props in rendering.
[mobx-react-devtools](https://github.com/mobxjs/mobx-react-devtools)'s observer tree shows "reactive props" if the component access props in render function, so I believe my guess.
![image](https://user-images.githubusercontent.com/13010755/45086625-96bf2c80-b13e-11e8-8dc3-9dbb8ea9ddb8.png)

But unlike what I think, render function called even if I didn't use props in there.
In [this sample](https://stackblitz.com/edit/react-mobx-pure?file=index.js), ` LoggableCounter`'s props won't use its props in render. but it re-render when I change its props by `App`'s input box.

I guess observer component has to call render when its props change because the event has to call component's lifecycle. Is that right?

then, isn't there an awesome way to prevent re-render when the component function doesn't use props in render function? I know I can optimize this by split a component or make prop boxed-value. but I wonder if those are the best ways. For example, I think mobx-react can provide decorater like `observer` but somethings that makes problems are excluded from the decorated component's lifecycle.

**Comment by urugator** [comment]
> awesome way to prevent re-render

Even better - if we would know that `props` wasn't used we could immediately return false from `shouldComponentUpdate` skipping shallow equality check.

The problem is that mobx tracks only `render`, but `componentDidMount`, `componentDidUpdate`, `getDerivedStateFromProps` can depend on props as well - they can affect rendered output.
So we would have to use different technique to detect props access in these methods.
Currently we are patching `getter/setter`, however I think that wouldn't work with `getDerivedStateFromProps`, because (correct me if I am wrong) it runs before the getter/setter is called. 

But most importantly I think that the utilization is so limited it's not worth to bother ... it's effective only if `this.props` (whole object) is accessed conditionally or if some prop is passed accidentally... or am I missing something?

**Comment by HealGaren** [comment]
> it's not worth to bother

Let's call the component `MobxPureComponent` that excludes the props-dependent lifecycle.
(This is one example of a solution to my issue. There will be many other solutions.)

If `mobx-react` provides a decorator that makes `MobxPureComponent`, I can optimize props-based render by using `computed`. I think it's worth to bother.

I know `props`  is immutable, so if I change partial of props the component cannot react selectively.
But [other example](https://stackblitz.com/edit/react-mobx-computed) is ready:

In my example, parent renders `count` and child renders `Math.floor(count/2)`.
so child's render count should be roughly half of parent's render count when I optimized them well.

There's two counter example.
In `react props` example, parent has observable `count` variable and it is passed to child.
In `store props` example, store has props like component. It's `observable.ref` variable. It should be changed by calling `setProps` and `setProps` makes it immutable like props.

according to [mobx-react-devtools](https://github.com/mobxjs/mobx-react-devtools)'s observer tree,
render function of child component in `react props` example depends on `reactive props`.
In `store props` example, the render function depends on the store's `props variable`.

In both examples, render function of the child component depends `props`(or props-like thing). And I tried optimizing them in the same way. But `react props` example cannot be optimized by computed-value.

React's passing props and making pure component is easy to read, and is easy to divide work. It would be very sad if I could not optimize the "props" provided and always had to write a store or boxed-value.

Thanks.
